source/iterator/cdc.go

```diff
@@ -0,0 +1,411 @@
+
+	reader, err := params.Conn.VStream(ctx, tabletType, vgtid, filter, &vtgate.VStreamFlags{
+		MinimizeSkew: false,
+	})
+	if err != nil {
+		return fmt.Errorf("create vstream reader: %w", err)
+	}
+	c.reader = reader
+
+	return nil
+}
+
+// findAllShardsInKeyspace executes a vtctl's FindAllShardsInKeyspace command, parses and returns the result.
+func (c *cdc) findAllShardsInKeyspace(ctx context.Context, address, keyspace string) ([]*binlogdata.ShardGtid, error) {
+	vtctlClient, err := vtctlclient.New(address)
+	if err != nil {
+		return nil, fmt.Errorf("vtctlclient connect: %w", err)
+	}
+	defer vtctlClient.Close()
+
+	// for more details,
+	// see https://vitess.io/docs/14.0/reference/programs/vtctldclient/vtctldclient_findallshardsinkeyspace/
+	eventStream, err := vtctlClient.ExecuteVtctlCommand(
+		ctx, []string{findAllShardsInKeyspaceCommand, keyspace}, defaultVtctlCommandTimeout,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("execute FindAllShardsInKeyspace vtctl command: %w", err)
+	}
+
+	event, err := eventStream.Recv()
+	if err != nil {
+		return nil, fmt.Errorf("receive FindAllShardsInKeyspace vtctl command result: %w", err)
+	}
+
+	eventStr := event.GetValue()
+	if eventStr == "" {
+		return nil, ErrFindAllShardsInKeyspaceReturnedNothing
+	}
+
+	shardsInfo := make(map[string]any)
+	if err := json.Unmarshal([]byte(eventStr), &shardsInfo); err != nil {
+		return nil, fmt.Errorf("unmarshal shards info: %w", err)
+	}
+
+	shardGtids := make([]*binlogdata.ShardGtid, 0, len(shardsInfo))
+	for shard := range shardsInfo {
+		shardGtids = append(shardGtids, &binlogdata.ShardGtid{
+			Keyspace: keyspace,
+			Shard:    shard,
+			Gtid:     defaultInitialGtid,
+		})
+	}
+
+	return shardGtids, nil
+}
+
+// constructRuleFilter constructs an SQL query for the binlogdata.Filter.Rules.
+func (c *cdc) constructRuleFilter(table, orderingColumn string, columns []string) (string, error) {
+	selectDataset := goqu.Dialect("mysql").Select()
+
+	if len(columns) > 0 {
+		keyColumnPresent := false
+		// add one to the capacity to have a space for the keyColumn
+		// if it's not present in the columns list.
+		cols := make([]any, 0, len(columns)+1)
+		for i := 0; i < len(columns); i++ {
+			if columns[i] == c.keyColumn {
+				keyColumnPresent = true
+			}
+
+			cols = append(cols, columns[i])
+		}
+
+		if !keyColumnPresent {
+			cols = append(cols, c.keyColumn)
+		}
+
+		selectDataset = selectDataset.Select(cols...)
+	}
+
+	query, _, err := selectDataset.
+		From(table).
+		Order(goqu.C(orderingColumn).Asc()).
+		ToSQL()
+	if err != nil {
+		return "", fmt.Errorf("construct rule filter query: %w", err)
+	}
+
+	return query, nil
+}
+
+// listen listens for VStream events.
+// If the VStream encountered an error the method will send it to the errCh channel.
+// All the data are sent to the records channel.
+func (c *cdc) listen(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
```

Comment:
The context here comes from the connector's `Open()` method, so I'm understanding the code as: we'll be listening to changes until `Open`'s context is cancelled. I wouldn't agree that that condition is correct. 

If `Open`'s context is cancelled, i.e. while `Open` is still running, then we shouldn't start listening to changes. However, once we started listening to changes, we should listening to them only if there was an error or `Stop` was called.

Reply:
I've added a `stopCh` channel to the `cdc` in 0c25fda8773f694ae72e21e65366c3d7329a9a23. The channel receives a value within the `Stop` method, and the `listen` listens to it.